45_api_versioning.md

```diff
@@ -0,0 +1,162 @@
+
+We want to achieve the following goals:
+
+* Allow older versions of agents to talk to newer versions of server-side components (within reason)
+* Allow minor upgrades to APIs without requiring a new version (new fields with reasonable defaults, etc)
+* Allow minor API version upgrades to server-side components to be transparent to existing agents without requiring their upgrade.
+* Implement in a way that doesn't require lots of copy/pasted code or near duplicate code for different versions of the API
+
+### Non-Goals
+
+The following are not goals for this enhancement:
+
+* Support for previous versions of the API structure. This one time can be a clean break.
+* Perpetual support for all versions of the API
+
+## Proposal
+
+Versioning will exist for all components and we will follow a `$MAJOR.$MINOR` numbering scheme similar to [Semantic Versioning](https://semver.org/). Backwards compatibility will be supported for at least 2 major versions. Forward compatability will be supported within the same major version family. 
+
+
+
+### User Stories
+
+#### Story 1
+
+An organization has a fleet of machines deployed with a `keylime-agent` which knows about a specific version of the keylime APIs (v1.2). The fleet's operators want to upgrade the keylime server side components to a newer version of keylime which is running a newer version (v2.3). When the upgrade is complete on the server side, the existing agents in the fleet should be able to continue their attestations as before without any problem communicating with the server.
+
+#### Story 2
+
+An organization is in the process of changing from a python-based `keylime-agent` to a rust-based one. The python-based agent was using the v1.1 version of the API and the new rust-based agent is using v1.2. The keylime server-side components are using the v1.3. Machines that have their agent replaced should continue to be able to perform attestations.
+
+#### Story 3
+
+An organization has upgraded their machine images to a new version of the `keylime-agent` which uses the v1.5 version of the API. Their existing machines use a previous version of the agent which uses the v1.2 version. Their server-side components are currently supporting v1.4 of the API. They first upgrade their server-side components to the latest version which is now at v2.1 of the API. Then new machines with the v1.5 client can communicate with the server side and existing clients running v1.4 will also continue to operate without problems. The fleet should be able to operate in this "mixed" mode of different client versions without issue.
+
+### Notes/Constraints/Caveats
+
+* The currect API will be pegged at `v1.0` and when the feature is released all components will be updated together. This will be the last time that this will be required as future versions will allow the client and server-side components to upgrade independently. But we need this clean break from previous URL paths to give us this foundation for the future.
```

Comment:
```suggestion
* The current API will be pegged at `v1.0` and when the feature is released, all components will be updated together. This will be the last time that this is required as future versions will allow the client and server-side components to upgrade independently. But we need this clean break from previous URL paths to give us this foundation for the future.
```